### ISSUE TITLE
global flags

### ISSUE THREAD

**Issue by jrperritt** [jrperritt/rack#42]
I want to open this up to discussion as well. There are some flags that every command can provide (such as output format and auth credentials. Currently, those flags exist with all the others in a command (namely, anywhere after the last subcommand). I did it this way originally so that users wouldn't have to know exactly where to put certain flags within the command, but I can understand how some may like the idea of global flags.

**Comment by rgbkrk** [comment]
It sure seems like output and auth are both global output flags. I guess we need to find which services are counter to this.

**Comment by rgbkrk** [comment]
I'm guessing file download may be the offender, but that might be about how we style it.

**Comment by kenperkins** [comment]
Current global flags:

- `username`
- `apiKey`
- `region`
- `output` or `json`, `csv`, ...

What else?

**Comment by jrperritt** [comment]
`authURL` (naming speculative)
`fields` (https://github.com/jrperritt/rack/pull/39) will allow users to limit the fields returned in output

**Comment by kenperkins** [comment]
Why would they need to provide auth url?

**Comment by rgbkrk** [comment]
output can be plain/table as well.

**Comment by jrperritt** [comment]
To authenticate to a different endpoint. See here: https://github.com/jrperritt/rack/issues/15, though that may be an admin-only function.

**Comment by kenperkins** [comment]
Ah, non production. I think it's sensible for that to stay, it's not *really* admin only like impersonation.

**Comment by jrperritt** [comment]
Probably `--debug` also.

**Comment by jnoller** [comment]
Under "completeness" we need servers, networks and block storage before cloud files.

**Comment by jrperritt** [comment]
Closing this. I think we've settled on having global flags.

**Comment by kenperkins** [comment]
I'm revisiting this issue. 

I used the client a bit today, and I find myself being annoyed having to move the cursor so much when manipulating parameters that are currently global.

For example:

```
$ rack servers instance list
$ rack servers instance list --json
$ rack servers instance list --json region IAD
$ rack servers instance list --json region DFW
$ rack servers instance list --json region ORD
```

I expect folks to make successive calls like above frequently, and having to move all the way back to `rack` in the cursor location seems sub optimal.

I realize I'm a bit contrarian to my opinions above, but like I said, this came from using it today.

**Comment by kenperkins** [comment]
Further examples:

```
$ rack servers instance list --profile ken
$ rack servers instance list --profile ken-test
$ rack servers instance list --profile production
```

**Comment by rgbkrk** [comment]
Is it the `cli` tooling we're using in golang that prevents us from having global flags that can be dropped anywhere after `rack`?

**Comment by jnoller** [comment]
-1 global before service

**Comment by jrperritt** [comment]
I'm pretty sure we can make any orientation of flags work, but I think we should decide on only one. For example, output and auth flags should either have to go after `rack` or have to go somewhere after the last subcommand. I don't think we should allow both forms.

**Comment by rtgoodwin** [comment]
Countering the -1 with a "consider before service" (my votes count as +.5). Cognitively, the leading/prefix sets up the environment, the trailing/suffix configures the command. That said I had a 30 min discussion with Ken and recognize this is something that can be sussed out when more config file work is testable and more users are playing with it. Also I'm a supernova fanboy. Clearly.

**Comment by jrperritt** [comment]
Implemented in #108 to allow output and auth flags both before service and after the last subcommand.
